clarity/src/vm/analysis/type_checker/tests/mod.rs

```diff
@@ -1986,13 +1986,27 @@ fn test_options(#[case] version: ClarityVersion, #[case] epoch: StacksEpochId) {
          (+ (foo (bar 1)) 1)
          ";
 
-    assert!(match mem_type_check(contract).unwrap_err().err {
-        CheckErrors::TypeError(t1, t2) => {
-            t1 == TypeSignature::from_string("(optional bool)", version, epoch)
-                && t2 == TypeSignature::from_string("(optional int)", version, epoch)
-        }
-        _ => false,
-    });
+    if version < ClarityVersion::Clarity2 {
+        assert!(
+            match mem_run_analysis(contract, version, epoch).unwrap_err().err {
+                CheckErrors::TypeError(t1, t2) => {
+                    t1 == TypeSignature::from_string("(optional bool)", version, epoch)
+                        && t2 == TypeSignature::from_string("(optional int)", version, epoch)
+                }
+                _ => false,
+            }
+        );
+    } else {
```

Comment:
Interesting behavior change.  I can see the benefit of making it so that the type-checker is now more fine-grained in its errors by reporting the innermost type of the `optional` wrapper.  But, I'm curious what changes to type-checking happened that caused this behavior?  If we can't pinpoint these, then this code might contain a consensus bug.

Reply:
Thanks, that makes sense.  Do type-checks cost more in Clarity2 as a result?

Reply:
Clarity2 type checks will have the cost assessed when a trait is reached, but there is not currently a cost for descending a compound type with no traits. I'll try to figure out a sensible approach here.

Reply:
Because the compound type can only be nested so deeply, maybe it's fine to just offer the descent for free?  I'd be on board if you're convinced that it can't be abused to make a contract run for a long time (which I think the nesting limit prevents).